pkg/server/server.go

```diff
@@ -0,0 +1,159 @@
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"time"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	cmclient "github.com/jetstack/cert-manager/pkg/client/clientset/versioned/typed/certmanager/v1"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+	securityapi "istio.io/api/security/v1alpha1"
+	"istio.io/istio/security/pkg/server/ca/authenticate"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cert-manager/istio-csr/cmd/app/options"
+	"github.com/cert-manager/istio-csr/pkg/util"
+)
+
+const (
+	IdentitiesAnnotationKey = "istio.cert-manager.io/identities"
+)
+
+// Server is the implementation of the istio CreateCertificate service
+type Server struct {
+	log *logrus.Entry
+
+	client cmclient.CertificateRequestInterface
+	auther authenticate.Authenticator
+
+	issuerRef   cmmeta.ObjectReference
+	preserveCRs bool
+}
+
+func New(log *logrus.Entry, cmOptions *options.CertManagerOptions, kubeOptions *options.KubeOptions) *Server {
+	return &Server{
+		log:         log.WithField("module", "certificate_provider"),
+		client:      kubeOptions.CMClient,
+		auther:      kubeOptions.Auther,
+		issuerRef:   cmOptions.IssuerRef,
+		preserveCRs: cmOptions.PreserveCRs,
+	}
+}
+
+// Run is a blocking func that will run the client facing certificate service
+func (s *Server) Run(ctx context.Context, tlsConfig *tls.Config, listenAddress string) error {
+	// Setup the grpc server using the passed TLS config
+	creds := credentials.NewTLS(tlsConfig)
+	grpcServer := grpc.NewServer(grpc.Creds(creds))
+
+	// listen on the configured address
+	listener, err := net.Listen("tcp", listenAddress)
+	if err != nil {
+		return fmt.Errorf("failed to listen %s: %v", listenAddress, err)
+	}
+
+	// register certificate service grpc API
+	securityapi.RegisterIstioCertificateServiceServer(grpcServer, s)
+
+	// handle termination gracefully
+	go func() {
+		<-ctx.Done()
+		s.log.Info("shutting down grpc server")
+		grpcServer.GracefulStop()
+		s.log.Info("grpc server stopped")
+	}()
+
+	s.log.Infof("grpc serving on %s", listener.Addr())
+
+	return grpcServer.Serve(listener)
+}
+
+// CreateCertificate is the istio grpc API func, to authenticate, authorize,
+// and sign CSRs requests from istio clients.
+func (s *Server) CreateCertificate(ctx context.Context, icr *securityapi.IstioCertificateRequest) (*securityapi.IstioCertificateResponse, error) {
+	// authn incoming requests, and build concatenated identities for labelling
+	identities, ok := s.authRequest(ctx, []byte(icr.Csr))
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "request authenticate failure")
+	}
+
+	// Build cert-manager CertificateRequest based on the configured issuer
+	cr := &cmapi.CertificateRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			// Random non-conflicted name
+			GenerateName: "istio-",
```

Comment:
do we have some name from the request? Thinking about a user debugging seeing 100 CRs called istio

Reply:
Yeah, we have the requester spiffe identity, which is a part of the annotation (mostly for auditing).

These id's can get too long for names.